cardano-cli/src/Cardano/CLI/Shelley/Run/Key.hs

```diff
@@ -0,0 +1,124 @@
+module Cardano.CLI.Shelley.Run.Key
+  ( ShelleyKeyCmdError
+  , renderShelleyKeyCmdError
+  , runKeyCmd
+  ) where
+
+import           Cardano.Prelude
+
+import qualified Data.Text as Text
+import qualified Data.Text.IO as Text
+
+import           Control.Monad.Trans.Except (ExceptT)
+import           Control.Monad.Trans.Except.Extra (firstExceptT,
+                   handleIOExceptT, hoistEither, hoistMaybe, newExceptT)
+
+import           Cardano.Api.TextView (TextViewDescription (..))
+import           Cardano.Api.Typed
+
+import           Cardano.CLI.Byron.Key (ByronKeyFailure, CardanoEra (..),
+                   readEraSigningKey, renderByronKeyFailure)
+import           Cardano.CLI.Helpers
+import           Cardano.CLI.Shelley.Parsers (ITNKeyFile (..), KeyCmd (..),
+                   OutputFile (..), SigningKeyFile (..),
+                   VerificationKeyFile (..))
+
+import qualified Cardano.Crypto.DSIGN.Class as Crypto
+import qualified Cardano.Crypto.Signing as Byron.Crypto
+import qualified Cardano.Crypto.Wallet as Crypto.HD
+
+import qualified Shelley.Spec.Ledger.Keys as Shelley
+
+data ShelleyKeyCmdError
+  = ShelleyKeyCmdReadFileError !(FileError TextEnvelopeError)
+  | ShelleyKeyCmdWriteFileError !(FileError ())
+  | ShelleyKeyCmdByronKeyFailure !ByronKeyFailure
+  | ShelleyKeyCmdByronKeyParseError
+      !Text
+      -- ^ Text representation of the parse error. Unfortunately, the actual
+      -- error type isn't exported.
+  | ShelleyKeyCmdDeserialiseByronVerKeyError
```

Comment:
I feel like it'd be nicer to instead have something like `ShelleyKeyCmdCborDecoderError !DecoderError`.